### PR TITLE
fix exact year inconsistencies in human readable duration

### LINKED ISSUE #89960: Human readable duration formats exact year inconsistently
**What happened**:
When the duration is **exactly** 2 years or 3 years, human readable duration is returned as `2y0d` or `3y0d`.  But when the duration is exactly 4 years, 5 years, etc, the human readable duration is returned as `4y`, `5y`, etc.

**What you expected to happen**:
The format of an exact year should be consistent.  Remove `0d` when duration is an exact year:
* `2y0d` should be `2y`
* `3y0d` should be `3y`
* Check other years... 4, 5, 6, 7... not sure if they also have the problem.

**How to reproduce it (as minimally and precisely as possible)**:
Because this is time-dependent, it can be difficult to reproduce via the CLI, but you can see the problem illustrated in [this unit test](https://github.com/kubernetes/kubernetes/blob/15bb54c2d2c515ad155b90c2c2ca59eaa1cd3f65/staging/src/k8s.io/apimachinery/pkg/util/duration/duration_test.go#L77) which shows the current behavior:
```
		{d: 2 * 365 * 24 * time.Hour, want: "2y0d"},
		{d: 2*365*24*time.Hour + 23*time.Hour, want: "2y0d"},
		{d: 3 * 365 * 24 * time.Hour, want: "3y0d"},
		{d: 8*365*24*time.Hour - time.Millisecond, want: "7y364d"},
		{d: 8 * 365 * 24 * time.Hour, want: "8y"},
		{d: 8*365*24*time.Hour + 364*24*time.Hour, want: "8y"},
		{d: 9 * 365 * 24 * time.Hour, want: "9y"},
```

The `2y0d` and `3y0d` test cases should be changed to want `2y` and `3y`, and the code should be changed to make that pass.

**Anything else we need to know?**:
Note, this is not a problem when the duration is exactly 1 year because the human readable duration is still shown in days.  So nothing needs to change to handle the case of exactly 1 year.

**Environment**:
- Kubernetes version (use `kubectl version`):  1.18
- Cloud provider or hardware configuration:  n/a
- OS (e.g: `cat /etc/os-release`):  n/a
- Kernel (e.g. `uname -a`):  n/a
- Install tools:  n/a
- Network plugin and version (if this is a network-related bug):  n/a
- Others:  n/a

### PR DESCRIPTION
**Pull request by bitscuit**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Human readable duration formats for exact years (e.g. 2 years, 3 years, etc.) are not consistent. They are currently displayed as 2y0d, 3y0d, ..., 7y0d. 8 years and above showed only years. Other exact units of time did not display 0, e.g. 2h and not 2h0m. This PR address this inconsistency by making exact years display only the year, and no day(s).

**Which issue(s) this PR fixes**:

Fixes #89960

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by bitscuit** [comment]
I wasn't sure if there should be release notes for this PR, so I erred on the safe side and put `NONE`. This could be a user-facing change though as the human readable duration format has been updated. Looking for guidance on the right course of action.

**Comment by bitscuit** [comment]
Should I run the `/retest` command? The  `pull-kubernetes-e2e-kind-ipv6` job failed because it could not create a cluster. I looked through issues related to the `pull-kubernetes-e2e-kind-ipv6` job and did not see any that resembled the one this PR is experiencing.

**Review by brianpursley — APPROVED** [review]

**Comment by brianpursley** [comment]
> Should I run the `/retest` command? The `pull-kubernetes-e2e-kind-ipv6` job failed because it could not create a cluster. I looked through issues related to the `pull-kubernetes-e2e-kind-ipv6` job and did not see any that resembled the one this PR is experiencing.

@bitscuit Yes, it is fine to use the `/retest` command if the test "flakes" due to some random condition. The tests are not supposed to do that, but some of them do and will succeed if you re-run them. 

You don't need to do it now though because tanjunchen already did it.

**Comment by bitscuit** [comment]
@brianpursley could you comment with `/lgtm`, I think that's required to move onto the next step

**Comment by brianpursley** [comment]
@bitscuit ci robot wants you to assign @deads2k ... https://github.com/kubernetes/kubernetes/pull/90082#issuecomment-612518427

**Comment by bitscuit** [comment]
assign deads2k

File: staging/src/k8s.io/apimachinery/pkg/util/duration/duration_test.go [hunk 1]
```diff
@@ -74,9 +74,14 @@ func TestHumanDurationBoundaries(t *testing.T) {
 		{d: 8 * 24 * time.Hour, want: "8d"},
 		{d: 8*24*time.Hour + 23*time.Hour, want: "8d"},
 		{d: 2*365*24*time.Hour - time.Millisecond, want: "729d"},
-		{d: 2 * 365 * 24 * time.Hour, want: "2y0d"},
-		{d: 2*365*24*time.Hour + 23*time.Hour, want: "2y0d"},
-		{d: 3 * 365 * 24 * time.Hour, want: "3y0d"},
+		{d: 2 * 365 * 24 * time.Hour, want: "2y"},
+		{d: 2*365*24*time.Hour + 23*time.Hour, want: "2y"},
+		{d: 2*365*24*time.Hour + 23*time.Hour + 59*time.Minute, want: "2y"},
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/util/duration/duration_test.go:79]: Check both sides of the boundary condition?

```
{d: 2*365*24*time.Hour + 24*time.Hour - 1, want: "2y"},
{d: 2*365*24*time.Hour + 24*time.Hour, want: "2y1d"}
```

Comment (bitscuit) [reply staging/src/k8s.io/apimachinery/pkg/util/duration/duration_test.go:79]: okay, should I do the same for the 8y boundary?

Comment (bitscuit) [reply staging/src/k8s.io/apimachinery/pkg/util/duration/duration_test.go:79]: oh, never mind. I just realized 8y boundary already does it

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/util/duration/duration_test.go:79]: The point is to demonstrate exactly where the cutoff is by having a case for each value. Looks like the boundary is not where I thought it was, judging from the 8y case.

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/util/duration/duration_test.go:79]: Oh, I see, the 8y case is a different code path completely.

**Comment by lavalamp** [comment]
/lgtm
/approve

btw, I think it's a little weird that these numbers are floor() rather than being rounded.

**Comment by bitscuit** [comment]
> 
> 
> /lgtm
> /approve
> 
> btw, I think it's a little weird that these numbers are floor() rather than being rounded.

maybe there was a reason for it? Though I've only just started contributing, so I wouldn't really know

**Pull request merged**